hal/src/common/thumbv6m/sercom/v2/spi.rs

```diff
@@ -0,0 +1,1768 @@
+impl_dipo_dopo!((Pad3, NoneT, Pad1, Pad2): (3, 0));
+impl_dipo_dopo!((Pad3, Pad0, Pad1, NoneT): (3, 0));
+impl_dipo_dopo!((Pad3, NoneT, Pad1, NoneT): (3, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, Pad2): (3, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, NoneT): (3, 0));
+
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, Pad1): (3, 1));
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, NoneT): (3, 1));
+
+//impl_dipo_dopo!((Pad3, NoneT, Pad1, Pad2): (3, 2));
+//impl_dipo_dopo!((Pad3, NoneT, Pad1, NoneT): (3, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, Pad2): (3, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, NoneT): (3, 2));
+
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, Pad1): (3, 3));
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, NoneT): (3, 3));
+
+//=============================================================================
+// Pads
+//=============================================================================
+
+/// Encapsulate the set of pads for an SPI peripheral
+///
+/// This struct acts to encapsulate up to four [`Pad`]s for use with an SPI
+/// peripheral. All of the [`Pad`]s must share the same [`Sercom`]. The four
+/// type parameters `DI`, `DO`, `CK` and `SS` represent the respective DI, DO,
+/// SCK and SS [`Pad`] types.
+///
+/// Each pad in this struct is an [`OptionalPad`]. When first initialized, each
+/// pad is set to [`NoneT`]. To be accepted as a valid set of [`Pads`] by the
+/// [`Config`] struct, the [`Pads`] must implement [`DipoDopo`], which requires
+/// the `CK` pad and either the `DI` or `DO` pad to be [`SomePad`].
+///
+/// The [`Sercom`] type must be specified up front and is the same for each
+/// [`Pad`], but each [`Pad`] will have different [`PadNum`] and [`Map`] types.
+/// For SAMD11 and SAMD21 chips, the [`Map`] type is always a [`PinId`].
+///
+/// Individual pads are set using a builder-pattern API. The argument to each
+/// function is a GPIO [`Pin`]. The [`PinId`] can be extracted from the [`Pin`]
+/// type, so there is no need to manually specify the [`Map`] type. But you will
+/// need to specify the desired [`PadNum`]. Based on the [`Sercom`], [`PadNum`]
+/// and [`PinId`] types, the [`Pin`] will be converted to the corresponding
+/// [`Pad`] automatically.
+///
+/// The following example corresponds to a set of [`Pads`] with `DIPO = 0` and
+/// `DOPO = 2`. It is possible to create a set of [`Pads`] that does not
+/// implement [`DipoDopo`], but such [`Pads`] will not be accepted by
+/// [`Config`].
+///
+/// ```no_run
+/// # use atsamd_hal::target_device::Peripherals;
+/// # use atsamd_hal::gpio::v2::Pins;
+/// # use atsamd_hal::sercom::v2::Sercom0;
+/// # use atsamd_hal::sercom::v2::pads::{Pad0, Pad1, Pad2, Pad3};
+/// # use atsamd_hal::sercom::v2::spi;
+/// let mut peripherals = Peripherals::take().unwrap();
+/// let pins = Pins::new(peripherals.PORT);
+/// let pads = spi::Pads::<Sercom0>::new()
+///     .data_in::<Pad0, _>(pins.pb08)
+///     .data_out::<Pad3, _>(pins.pb11)
+///     .sclk::<Pad1, _>(pins.pb09);
+/// ```
```

Comment:
It might be useful as well to add a note that you need to provide the pins as `v2` GPIO types, to avoid needing verbose type declarations.

Reply:
I thought I designed it to work with either pin type. Is that not working for you? See my comment on Matrix

Reply:
Nop, due to ambiguity in type inference :/ More details in my comment on matrix

Reply:
With into:

```
cannot infer type for struct `Pad<SERCOM7, Pad1, IoSet4>`
```

Without:
```
the trait `From<atsamd_hal::gpio::Pin<PB20, atsamd_hal::gpio::v2::Input<atsamd_hal::gpio::v2::Floating>>>` is not implemented for `Pad<SERCOM7, Pad1, IoSet4>
```

Reply:
I should have a fix for this.

Reply:
I can confirm this works with `v1` pins with your Wio Terminal diff